### PR TITLE
Expect exit code enhancement

### PR DESCRIPTION
**Pull request by tjungblu**
ExpectProcess and ExpectFunc now take the exit code of the process into account, not just the matching of the tty output.

Signed-off-by: Thomas Jungblut <tjungblu@redhat.com>

I'm still fixing some of the tests that are now failing due to this.

### CODE REVIEW

**Comment by serathius** [comment]
Nice, one additional option to consider is returning process status so the caller can decide based on exit code. Example linearizability tests expect process to fail when tailpoint is injected.

**Comment by tjungblu** [comment]
added the status code as well, I think this is going to become a bigger refactor since we there was no exit code assumption beforehand. Some tests still don't run, I also think there's some locking issue. Gotta continue tomorrow.

File: tests/e2e/ctl_v3_auth_test.go
Comment (tjungblu) [inline tests/e2e/ctl_v3_auth_test.go:683]: sorry this was some autocorrect (wtf)

**Comment by ahrtr** [comment]
Thanks for working on this. 

It seems the PR caused lots of workflow failures. Usually 1~2 failures might be caused by flaky test. But 6 failures most likely mean that it's caused by the PR.  Please take a look at the failures.

File: pkg/expect/expect.go
Comment (tjungblu) [inline pkg/expect/expect.go:245]: this is just temporary, a compile error regarding Wait() came out of nowhere, just interested what was trying to consume this. There's no interface around it...

Comment (tjungblu) [reply pkg/expect/expect.go:245]: Turned out to be a change in main, I didn't know the action would apply this commit on top of main and run the tests on top of it. I've rebased and could fix it.

**Comment by tjungblu** [comment]
@ahrtr @serathius finally managed to get all to green. There are a couple of weird choices I've made around the error assertions, so let me know what you think about it.

File: pkg/expect/expect.go
Comment (serathius) [inline pkg/expect/expect.go:105]: no need for anonymous function

Comment (tjungblu) [reply pkg/expect/expect.go:105]: disagree, this makes the locking more obvious in 

```
ep.mu.Lock()
defer ep.mu.Unlock()
```

File: tests/e2e/ctl_v3_auth_test.go
Comment (tjungblu) [inline tests/e2e/ctl_v3_auth_test.go:681]: this should also be a require, I dunno why I picked assert here instead

File: pkg/expect/expect.go
Comment (serathius) [inline pkg/expect/expect.go:226]: this is weird decision. would prefer to return an error or some known error code like 255

Comment (tjungblu) [reply pkg/expect/expect.go:226]: agreed, will change this

File: tests/common/defrag_test.go
Comment (serathius) [inline tests/common/defrag_test.go:50]: Why defragment command fails here?

File: tests/e2e/cluster_downgrade_test.go
Comment (serathius) [inline tests/e2e/cluster_downgrade_test.go:51]: This looks like unrelated changes from other PR. Please rebase.

File: tests/e2e/ctl_v3_auth_test.go
Comment (serathius) [inline tests/e2e/ctl_v3_auth_test.go:124]: Nice!

File: tests/common/defrag_test.go
Comment (tjungblu) [reply tests/common/defrag_test.go:50]: odd, works again. Reverted that.

File: tests/e2e/cluster_downgrade_test.go
Comment (tjungblu) [reply tests/e2e/cluster_downgrade_test.go:51]: I can split this into another commit, but it was never passing for me locally. Not sure if that works on your box.

File: tests/e2e/ctl_v3_elect_test.go
Comment (serathius) [inline tests/e2e/ctl_v3_elect_test.go:103]: I would prefer to handle failure by caller.

File: tests/e2e/ctl_v3_move_leader_test.go
Comment (serathius) [inline tests/e2e/ctl_v3_move_leader_test.go:116]:
```suggestion
		expectError bool
```

File: tests/e2e/ctl_v3_elect_test.go
Comment (tjungblu) [reply tests/e2e/ctl_v3_elect_test.go:103]: me too, do you have an idea how I can do this from the go routine down there? Otherwise I'd have to add another error channel

File: tests/e2e/ctl_v3_move_leader_test.go
Comment (tjungblu) [reply tests/e2e/ctl_v3_move_leader_test.go:116]: updated

File: tests/framework/e2e/etcd_process.go
Comment (serathius) [inline tests/framework/e2e/etcd_process.go:183]: This totally changes the result. Don't know any case we would want to change it like this.

Comment (tjungblu) [reply tests/framework/e2e/etcd_process.go:183]: good call, it's only used in the failpoints. Seems to work though,

**Comment by serathius** [comment]
For commands that fail authorization it makes sense to introduce `require.ErrorContains(cx.t, err, "permission denied")`, however I don't understand why we change `t.Fatal(err)` to require.ErrorContains(cx.t, err, "unexpected exit code").

**Comment by tjungblu** [comment]
Let me go through all of them again and replace them with something more meaningful. There were a bunch of cases where the process was killed and the return was non-zero with no meaningful output.

File: tests/e2e/ctl_v3_auth_test.go
Comment (tjungblu) [inline tests/e2e/ctl_v3_auth_test.go:982]: @serathius here is one of such cases where no meaningful error output comes due to the watch being cancelled:

>     ctl_v3_auth_test.go:982: [unexpected exit code [5] after running [/home/tjungblu/git/etcd/bin/etcdctl --endpoints=http://localhost:20000 --dial-timeout=7s --user=test-user:pass watch key5 --rev 1]]

**Comment by serathius** [comment]
>  There were a bunch of cases where the process was killed and the return was non-zero with no meaningful output.

It should be clear on caller side that the process was killed and is expected to exit with specific exit code. Processes killed by signal should return specific exit codes.

File: tests/e2e/ctl_v3_elect_test.go
Comment (tjungblu) [inline tests/e2e/ctl_v3_elect_test.go:80]: @serathius that's another candidate:
```
[unexpected exit code [1] after running [/home/tjungblu/git/etcd/bin/etcdctl --endpoints=http://localhost:20000 --dial-timeout=7s elect a p2]]
```

Comment (tjungblu) [reply tests/e2e/ctl_v3_elect_test.go:80]: actually `CloseWithTimeout` also looks a little more difficult than it needs to be.

**Comment by tjungblu** [comment]
@serathius there are two odd cases where it could either be canceled, killed or successful using `CloseWithTimeout` on interrupted processes (`testElect` + `testLock`):

https://github.com/etcd-io/etcd/blob/7ed4eda4c19ca2f1eeb461d7abb7aee6a244f7a7/tests/framework/e2e/util.go#L112-L124

I've specifically commented those cases, but let me know if you think there's a better way to handle this. The rest should be cleaned up properly now.

TBH I'm also not entirely sure what these tests are supposed to test?
Introduced in https://github.com/etcd-io/etcd/pull/5059, kinda duped to https://github.com/etcd-io/etcd/pull/5056

File: pkg/expect/expect.go
Comment (serathius) [reply pkg/expect/expect.go:108]: Sounds good.

File: tests/e2e/ctl_v3_auth_test.go
Comment (serathius) [reply tests/e2e/ctl_v3_auth_test.go:981]: Makes sense.

File: tests/e2e/ctl_v3_elect_test.go
Comment (serathius) [reply tests/e2e/ctl_v3_elect_test.go:104]: Seems ok for now.

File: tests/framework/e2e/etcd_process.go
Comment (serathius) [inline tests/framework/e2e/etcd_process.go:104]: Why remove this?

File: tests/framework/e2e/etcd_process.go
Comment (serathius) [reply tests/framework/e2e/etcd_process.go:183]: It might work, but it causes a race between previously sent SIGKILL signal and process close. Please revert.

File: tests/framework/e2e/util.go
Comment (serathius) [inline tests/framework/e2e/util.go:53]: There is no expectation that this commands should only take 2 seconds. I expect because this function is mostly used for running `etcdctl` it work, but the the name of the function do not suggest limited run time. 

If limiting run time is required for tests to pass, please move it up the call stack.

File: tests/e2e/cluster_downgrade_test.go
Comment (serathius) [reply tests/e2e/cluster_downgrade_test.go:51]: I expect that after rebase those changes will not be needed/present.

File: pkg/expect/expect.go
Comment (serathius) [inline pkg/expect/expect.go:78]: Waiting for `waitSaveExitErr` changes the behavior of close, now it will always wait till process exit. I think we should keep `Wait` and `Close` separate.
* `Close` should be used to ensure cleanup, even if process doesn't close by itself. To ensure process exists by itself it should send SIGTERM and wait couple of seconds for process to exit. Process should be expected to exist with success.
* `Wait` should expect the process to exit by itself and just wait till that moment.

File: tests/framework/e2e/etcd_process.go
Comment (serathius) [reply tests/framework/e2e/etcd_process.go:183]: Reason is that you changed implementation of `Close` to always wait till the process finishes. This however not correct behavior.

File: tests/framework/e2e/util.go
Comment (serathius) [inline tests/framework/e2e/util.go:91]: We should wait for process to exit by itself.

**Comment by serathius** [comment]
Great work overall. I expect this will be crucial to reduce flakiness of e2e tests. One think I would like to fix before we merge is separation between `Wait` and `Close`. They should behave like their process level counterparts.

File: tests/e2e/cluster_downgrade_test.go
Comment (tjungblu) [reply tests/e2e/cluster_downgrade_test.go:51]: I don't understand, the version inconsistency I run into still exists in main:
https://github.com/etcd-io/etcd/blob/main/tests/e2e/cluster_downgrade_test.go#L50-L55

The issue stems from the fact that the logic takes what is assumed to be in the binary, not what is actually reported by it. If you check the diff, it's already against the latest revision of the test.

File: pkg/expect/expect.go
Comment (tjungblu) [reply pkg/expect/expect.go:78]: So you always have to run `Close` after `Wait`? You can run `close` without `wait` however.

I'll check this out, but I'm sure we have to rewrite a few more tests for this.

File: tests/framework/e2e/util.go
Comment (tjungblu) [reply tests/framework/e2e/util.go]: removed again

File: tests/framework/e2e/etcd_process.go
Comment (tjungblu) [reply tests/framework/e2e/etcd_process.go]: added it back

File: tests/e2e/cluster_downgrade_test.go
Comment (serathius) [reply tests/e2e/cluster_downgrade_test.go:51]: Can you create a separate PR to fix the downgrade test?

File: pkg/expect/expect.go
Comment (serathius) [reply pkg/expect/expect.go:78]: > So you always have to run Close after Wait?

Don't think so, process exits if called by any of them. The main difference is that Wait wait for process to exit normally and Close should always guarantee the exit by sending termination signal.

Comment (tjungblu) [reply pkg/expect/expect.go:78]: hmm. What do you think about the following semantics:
* `Wait` just waits for the process to exit
* `Close` waits for the process to exit + cleanup (required call)
* `Stop` signals SIGTERM to the process

> Close should always guarantee the exit by sending termination signal.

So `Close` should always send SIGTERM?

File: pkg/expect/expect.go
Comment (tjungblu) [reply pkg/expect/expect.go:281]: I was a bit on the fence when I moved to a slice of errors here, the reason is that we shouldn't really be hiding any errors as this makes CI debugging a lot more difficult. 
As you can see I ended up filtering all relevant io error messages anyway. Next commit will go back to a single error.

File: tests/e2e/cluster_downgrade_test.go
Comment (tjungblu) [reply tests/e2e/cluster_downgrade_test.go:51]: sure, I'll pull this out of that PR.

Comment (tjungblu) [reply tests/e2e/cluster_downgrade_test.go:51]: created it: https://github.com/etcd-io/etcd/pull/14710

File: pkg/expect/expect.go
Comment (serathius) [reply pkg/expect/expect.go:78]: I think `Close` should work like standard linux process close procedure. You first ask nicely for process to exit by sending `SIGTERM` and then wait until process exits. If it doesn't exit within allotted time (e.g. 10s), you kill it using `SIGKILL`.

File: pkg/expect/expect.go
Comment (serathius) [inline pkg/expect/expect.go:295]: I think it might be useful to at least log the error.

File: pkg/expect/expect.go
Comment (serathius) [reply pkg/expect/expect.go:78]: Feel free to disregard by comment. Looking at current implementation of `etcd_process.go` I think what you proposed is ok.

**Review by serathius — APPROVED** [review]
LGTM, this is a big change so lets get also feedback from other maintainers.
cc @ahrtr @spzala @ptabor

**Comment by tjungblu** [comment]
Thanks Marek, I'm still fixing the remainder of the tests that are failing due to the changes

File: pkg/expect/expect.go
Comment (tjungblu) [reply pkg/expect/expect.go:295]: added

File: pkg/expect/expect.go
Comment (tjungblu) [reply pkg/expect/expect.go:78]: okay cool, I think we should still add what you proposed to `close` as there is a risk it can hang waiting. It could be a good follow-up PR.

Comment (serathius) [reply pkg/expect/expect.go:78]: sg

**Comment by serathius** [comment]
> oh wow, almost all green. 

Great job!

> The last check seems to be an unrelated flake:

Rerun the failed scenario

> I'm just testing whether the e2e test got a lot longer than before, otherwise this is good to go from my side.

Correctness here is more important then time. Let's revisit it when time becomes a problem.

**Comment by tjungblu** [comment]
Thanks for the rerun. It's almost the same runtime on e2e on my local box. I'll keep an eye on the execution times on the workflow over the next couple days.

File: pkg/expect/expect.go
Comment (ahrtr) [inline pkg/expect/expect.go:142]:
```suggestion
	return err
```

File: tests/framework/e2e/cluster.go
Comment (ahrtr) [inline tests/framework/e2e/cluster.go:653]: Minor comment: if there is no parameter, then you'd better use `errors.New` instead.

Comment (tjungblu) [reply tests/framework/e2e/cluster.go:653]: fixed

File: pkg/expect/expect.go
Comment (tjungblu) [reply pkg/expect/expect.go]: fixed

**Comment by ahrtr** [comment]
This PR is based on an old commit https://github.com/tjungblu/etcd/commit/0d4a516d11d933dc15388fe64b760820ab3d77fc, please rebase this PR, then I will take another look.

**Comment by tjungblu** [comment]
@ahrtr just rebased, PTAL.

**Review by ahrtr — APPROVED** [review]
Overall looks good to me. 

Great work. Thank you @tjungblu

**Comment by tjungblu** [comment]
Thanks @ahrtr. Please ping me on any flakes and perf regressions, I'll monitor this throughout the week.

**Pull request merged**
